### ISSUE TITLE
Cannot use ivy backend

### ISSUE THREAD

**Issue by sriramkswamy** [jkitchin/org-ref#214]
I have set the following as per README in my `init.el` but it still uses `helm-bibtex` for citation completion, `helm` for navigating options of search results from `crossref-add-bibtex-entry` and so on. It also uses `ido` to select the correct `.bib` file for input and not `ivy`. 

    (setq org-ref-completion-library 'org-ref-ivy-bibtex)

I'm running OS X 10.11.4 and tested it with both the Emacs 25.1 pretest version and the Emacs Mac port, both compiled from homebrew.

    GNU Emacs 25.1.50.1 (x86_64-apple-darwin15.4.0, NS appkit-1404.46 Version 10.11.4 (Build 15E65)) of 2016-04-28

    GNU Emacs 24.5.1 (x86_64-apple-darwin15.4.0, Carbon Version 157 AppKit 1404.46) of 2016-05-05 on Srirams-MacBook-Pro.local

This is my relevant section of the configuration -

    (require 'org-ref)
    (require 'org-ref-pdf)
    (require 'org-ref-url-utils)
    (setq org-ref-completion-library 'org-ref-ivy-bibtex)
    (setq org-ref-notes-directory "~/Dropbox/org/references/notes"
          org-ref-bibliography-notes "~/Dropbox/org/references/articles.org"
          org-ref-default-bibliography '("~/Dropbox/org/references/multiphysics.bib" "~/Dropbox/org/references/chanceconstraints.bib")
          org-ref-pdf-directory "~/Dropbox/org/references/pdfs/")

Edit: The previous issue on downloading PDFs was resolved. I had uncommented the `(require 'org-ref-pdf)` part in my config. Requiring it explicitly solved the issue.

**Comment by jkitchin** [comment]
The ivy integration is not complete. crossref-add-bibtex-entry for example is coded with a helm menu. 

The ivy backend should work for inserting citations, but I wouldn't expect it to change everything over to ivy completion. I am not totally sure how to do that either. That would be nice though.

**Comment by ericdanan** [comment]
It shouldn't use `helm-bibtex` for citation completion. I think the problem is that you need to put

```
(setq org-ref-completion-library 'org-ref-ivy-bibtex)
```

before

```
(require 'org-ref)
```

in your `init.el`.

But `crossref-add-bibtex-entry` indeed uses `helm` and `ido` in any case.

**Comment by sriramkswamy** [comment]
@ericdanan I changed it as per your suggestion and it didn't make any difference.

With respect to ivy completion instead of ido, it seems using `completing-read` instead of `ido-completing-read` uses ivy since I have set `(ivy-mode 1)`. I expect that if helm is turned on with `(helm-mode 1)`, it should use helm but it might hamper the out-of-the-box user experience because if someone doesn't use either, they might get the default completion. However, I think this problem can be solved by defaulting to ido when both are unavailable. This, after removing the requirement of helm as suggested in #213, should use the preferred completion.

With respect to changing the helm completion, I followed these posts ([1](https://www.reddit.com/r/emacs/comments/46ibsx/why_i_use_helm_and_why_you_should_too/d05pdow) and [2](http://oremacs.com/2015/07/09/counsel-rhythmbox/)) from the author of ivy and fiddled around the source but I'm an amateur at elisp and hence didn't get very far. Basically, the author encourages to use `completing-read` API instead of using helm directly. I'll keep fiddling around with it, but in the meanwhile it might give you others an idea as to how to approach this problem.

**Comment by ericdanan** [comment]
I don't know what the problem is then. Which function do you use to complete citations? What happens if after loading `org-ref` you do `M-x org-ref-change-completion` and select `org-ref-ivy-bibtex`?

**Comment by jkitchin** [comment]
I am not able to reproduce your issue. With the files at https://gist.github.com/jkitchin/4e96bca71ba1dd2dc5e82290063182c4

I can run "make ivy" and it does what I expect. (after "make new" which installs everything into the sandbox.)

**Comment by sriramkswamy** [comment]
@ericdanan The only option I seem to have for `org-ref-change-completion` is `org-ref-helm-cite`. I have ivy-bibtex installed and I directly invoke it by using `M-x ivy-bibtex`. It works well.

@jkitchin I cloned the set of files and then ran `make new`, quit emacs and then ran `make ivy` but trying to do `org-ref-cite-with-completion` invokes ido, not ivy, and goes to the default completion method after I have chosen the citation style. There are separate functions for `org-ref-helm-cite*` but that (obviously) invokes helm without any problem. Am I using the correct functions? I didn't see any function like `org-ref-ivy*`. I also tried changing the completion interactively by invoking `M-x org-ref-change-completion` and I still got `org-ref-helm-cite` to be the sole completion.

**Comment by ericdanan** [comment]
I guess that in `org-ref-change-completion` you get `org-ref-helm-cite` as
sole completion because it is the default candidate and is pre-inserted in
the minibuffer. If you delete it from the minibuffer you should see the
other candidates as well.

There is a dedicated function `org-ref-ivy-insert-cite-link` to insert a
citation with `ivy` completion. It is defined in file
`org-ref-ivy-bibtex.el` from the `org-ref` package, and is independent of
`ivy-bibtex.el` from the `bibtex-completion` / `ivy-bibtex` package.

**Comment by sriramkswamy** [comment]
> I guess that in `org-ref-change-completion` you get `org-ref-helm-cite` as
sole completion because it is the default candidate and is pre-inserted in
the mini buffer.

Yes, I deleted it and it worked.

I was about to make an edit that I found the `org-ref-ivy*` versions but you beat me to it.

I'm closing this issue for now. Probably in the future someone (or myself when I learn more elisp) will contribute and make sure all completions (like `crossref-add-bibtex-entry`) work with ivy/any other framework.
